### ISSUE TITLE
TypeScript compiler crashes duktape

### ISSUE THREAD

**Issue by amol-** [svaarala/duktape#449]
Trying to compile typescriptServices.js ( from https://rawgit.com/Microsoft/TypeScript/master/lib/typescriptServices.js ) leads to a duktape crash:

```
$ ./duk ../../dukpy/typescriptServices.js 
Segmentation fault: 11
```

Error during crash:

```
* thread #1: tid = 0x70a0f, 0x000000010002b940 duk`duk_js_execute_bytecode + 1280, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x000000010002b940 duk`duk_js_execute_bytecode + 1280
duk`duk_js_execute_bytecode:
->  0x10002b940 <+1280>: movq   0x8(%rax), %rcx
    0x10002b944 <+1284>: addq   $0x1, %rcx
    0x10002b94b <+1291>: movq   %rcx, 0x8(%rax)
    0x10002b94f <+1295>: jmp    0x10002b954               ; <+1300>
```

From time to time it doesn't crash but fails with:

```
TypeError: not number
	duk_api_stack.c:1157
	anon ../../dukpy/typescriptServices.js:140
	anon ../../dukpy/typescriptServices.js:349
	global ../../dukpy/typescriptServices.js:831 preventsyield
error in executing file ../../dukpy/typescriptServices.js
```

**Comment by svaarala** [comment]
Could you try what happens with assertions and self tests enabled?

```
-DDUK_OPT_ASSERTIONS -DDUK_OPT_SELF_TESTS
```

**Comment by JoshEngebretson** [comment]
I've wanted to try this for awhile and would be really curious on timings for compiling TypeScript under duktape

**Comment by amol-** [comment]
With ```-DDUK_OPT_ASSERTIONS -DDUK_OPT_SELF_TESTS``` I get the following error:

```
PANIC 54: assertion failed: DUK_TVAL_IS_NUMBER(tv) (duk_api_stack.c:2244) (calling abort)
Abort trap: 6
```

instead of the traceback when it doesn't crash

and i get the following when it crashes

``` 
Process 56349 stopped
* thread #1: tid = 0x7bfe2, 0x000000010004e8cc duk`duk_js_execute_bytecode + 2988, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x000000010004e8cc duk`duk_js_execute_bytecode + 2988
duk`duk_js_execute_bytecode:
->  0x10004e8cc <+2988>: movl   (%rax), %ecx
    0x10004e8ce <+2990>: movl   %ecx, %eax
    0x10004e8d0 <+2992>: andq   $0x3, %rax
    0x10004e8d6 <+2998>: cmpq   $0x1, %rax
```

**Comment by svaarala** [comment]
This is the switch-case which causes the assert (I'm assuming you're running 1.3.0):

- https://github.com/svaarala/duktape/blob/v1.3.0/src/duk_api_stack.c#L2185-L2252

That switch-case should be complete, i.e. only numbers should fall into the default clause. The most likely cause for this is either memory corruption somewhere, or some Duktape bug in initializing a tagged value somewhere.

**Comment by svaarala** [comment]
This is what I get:

```
$ ./duk.O2.130 typescriptServices.js 
SyntaxError: invalid regexp quantifier (unknown char) (line 1193)
	typescriptServices.js:1193
	duk_lexer.c:1659
error in executing file typescriptServices.js
```

How are you compiling the problem file - are you eval'ing it from an application or something similar? If so, could you try if the problem also happens with the Duktape command line tool?

**Comment by svaarala** [comment]
Ok, fixing that regexp manually, I get a segfault too, so seems to be reproducible.

**Comment by svaarala** [comment]
There's no segfault in Duktape master, so I should be able to bisect where this got fixed.

**Comment by amol-** [comment]
Yep, sorry, forgot to mention that I manually fixed the regex to make it compile.

**Comment by svaarala** [comment]
Needed to a [reverse bisect](http://melix.github.io/blog/2014/07/bisecting-groovy.html). The commit which fixes this issue between v1.3.0 and master is 5ae7ab87435140e5f5549429fa71f502f5a13804. The preceding "bad" commit is 8966c67c937cea5f4197882fb1754fe2a57c3222.

I think this will be enough to fix the issue and include a fix into 1.3.1. I'll look into this tomorrow - thanks for reporting! :)

If you want to use the master in the meanwhile, there are dist snapshots in duktape.org/snapshots, e.g. http://duktape.org/snapshots/duktape-1.3.99-20151121020005-v1.3.0-268-gbcdee88.tar.xz.

**Comment by svaarala** [comment]
I've labeled this "security" because there's an unintentional segfault.

**Comment by fatcerberus** [comment]
I suspected it might have been related to the recent temp fixes once you said it was already fixed in master.

**Comment by svaarala** [comment]
Added the following asserts to 1.3.0 release:

```
$ git diff
diff --git a/src/duk_js_compiler.c b/src/duk_js_compiler.c
index 6905c31..3a09558 100644
--- a/src/duk_js_compiler.c
+++ b/src/duk_js_compiler.c
@@ -2055,12 +2055,14 @@ duk_regconst_t duk__ispec_toregconst_raw(duk_compiler_ctx *comp_ctx,
                } else {
                        if (forced_reg >= 0) {
                                if (x->regconst != (duk_regconst_t) forced_reg) {
+                                       DUK_ASSERT((x->regconst & DUK__CONST_MARKER) == 0);
                                        duk__emit_a_bc(comp_ctx, DUK_OP_LDREG, forced_reg, x->regconst);
                                }
                                return (duk_regconst_t) forced_reg;
                        } else {
                                if ((flags & DUK__IVAL_FLAG_REQUIRE_TEMP) && !DUK__ISTEMP(comp_ctx, x->regconst)) {
                                        duk_reg_t dest = DUK__ALLOCTEMP(comp_ctx);
+                                       DUK_ASSERT((x->regconst & DUK__CONST_MARKER) == 0);
                                        duk__emit_a_bc(comp_ctx, DUK_OP_LDREG, (duk_regconst_t) dest, x->regconst);
                                        return (duk_regconst_t) dest;
                                } else {
```

When running with assertions enabled:

```
PANIC 54: assertion failed: (x->regconst & DUK__CONST_MARKER) == 0 (duk_js_compiler.c:2065) (segfaulting on purpose)
```

This is the second added assert. In short, the "intermediate value" handling code in 1.3.0 has a bug when (1) the call site requests a temporary register explicitly, and (2) the existing ivalue is a constant. In this case an LDREG will be emitted, with the LDREG referencing a constant, which results in a garbage value being loaded at runtime.

As noted above, this has been fixed in master, and the `v1.3-maintenance` branch also contains the fix already, so that it's already in place for a 1.3.1 release.

**Comment by svaarala** [comment]
Fix is also already in v1.2-maintenance for the 1.2.5 release (last in 1.2 I think).
